### ISSUE TITLE
wrong tensor arrange for conv

### ISSUE THREAD

**Issue by inspirit** [lucidrains/tranception-pytorch#1]
Hey!
looks like there is a mess in tensors arrangements in few place here:
https://github.com/lucidrains/tranception-pytorch/blob/610ebf21c4cf3b7f0d560c89610051c7242527a0/tranception_pytorch/tranception_pytorch.py#L140
does not make sense to go to 'b h n d' last dim is channels dim

than next
https://github.com/lucidrains/tranception-pytorch/blob/610ebf21c4cf3b7f0d560c89610051c7242527a0/tranception_pytorch/tranception_pytorch.py#L148

we split heads into groups and merge with batch and try to apply convolution that expects channels as second dim which is seq-len in our case now.

there is also 
https://github.com/lucidrains/tranception-pytorch/blob/610ebf21c4cf3b7f0d560c89610051c7242527a0/tranception_pytorch/tranception_pytorch.py#L122
setup of convolution layer to expect full inner_dim as input channels somehow

**Comment by inspirit** [comment]
grouped_sims never used actually :)
https://github.com/lucidrains/tranception-pytorch/blob/610ebf21c4cf3b7f0d560c89610051c7242527a0/tranception_pytorch/tranception_pytorch.py#L165

also Alibi impl here has an issue with backward call complaining about missing/overwritten gradient, after replacing it with another impl from x-transformers repo it works fine.

causal d-convs should work on `dim_head` instead of `inner_dim` or may be on `(heads//groups) * dim_head` its not clear from the paper, and if we refer to Primer impl they seem to run d-convs after projecting to `q-k-v` but before splitting to heads...

**Comment by lucidrains** [comment]
@inspirit oh hey! haha, this repository was far from ready, but thanks for giving it an early review

https://github.com/lucidrains/tranception-pytorch/commit/d3944ef4566a5a7d36d02564e5b2a8b6c474cc56 do you want to see if most (all?) the issues were addressed?

**Comment by lucidrains** [comment]
i wasn't able to replicate the ALiBi error, so if you have a short script i can run, i can definitely fix that as well!

**Comment by inspirit** [comment]
It seems the issue with ALiBi is due to few network forward calls before backward call, specifically i have 
```
p0 = net(x)
with torch.no_grad():
    p = net(y)
loss = ....
loss.backward()
```

**Comment by inspirit** [comment]
mask needs to be moved to correct device here:
https://github.com/lucidrains/tranception-pytorch/blob/a09cba2a66952ce4fb0f57f8441cba0b1eb15d37/tranception_pytorch/tranception_pytorch.py#L177

i also think its good to have it optional in case we dont need causual masking?

**Comment by lucidrains** [comment]
> mask needs to be moved to correct device here:
> 
> https://github.com/lucidrains/tranception-pytorch/blob/a09cba2a66952ce4fb0f57f8441cba0b1eb15d37/tranception_pytorch/tranception_pytorch.py#L177
> 
> i also think its good to have it optional in case we dont need causual masking?

sure! done in the latest!

so i'm still not seeing the ALiBi error

below is what i'm running

```python
import torch
from tranception_pytorch import Tranception

model = Tranception(
    dim = 512,
    depth = 6,
    heads = 8,
    dim_head = 64
)

amino_acids = torch.randint(0, 21, (1, 512))

logits = model(amino_acids)

with torch.no_grad():
    _ = model(amino_acids)
    _ = model(amino_acids)
    _ = model(amino_acids)

logits.sum().backward()
```

**Comment by inspirit** [comment]
i noticed that you apply alibi here inside forward:
https://github.com/lucidrains/tranception-pytorch/blob/128400a53e69e9691c9c593e38233d20a353aee3/tranception_pytorch/tranception_pytorch.py#L52

however the interface of function assume it will not apply bias but will return it instead for manual addition in attention class:
https://github.com/lucidrains/tranception-pytorch/blob/128400a53e69e9691c9c593e38233d20a353aee3/tranception_pytorch/tranception_pytorch.py#L61

and here we follow this idea:
https://github.com/lucidrains/tranception-pytorch/blob/128400a53e69e9691c9c593e38233d20a353aee3/tranception_pytorch/tranception_pytorch.py#L174

manually adding returned bias

Update: yup, that seems to be an issue, need to return bias or dont accumulate it inside attention class

**Comment by lucidrains** [comment]
@inspirit thank you Eugene! https://github.com/lucidrains/tranception-pytorch/commit/bc5002498b88ab84035e2fabecf962e1f45d2dbe

**Comment by lucidrains** [comment]
@inspirit the repository is still missing the retrieved MSA's contribution to the prediction but i'll get to that later next week!
